lua_modules/sntp/sntp.lua

```diff
@@ -0,0 +1,226 @@
+    end
+
+    return res and sntppkt.read_resp(res), best
+  end
+
+  local nextServer
+  local doserver
+
+  -- Try communicating with the current server
+  local function hail(ip)
+    print("sntp", "hail", ip)
+
+    _tmr:alarm(SNTP_TIMEOUT, tmr.ALARM_SINGLE, function()
+      print("sntp", "hail-tmr")
+      _udp:on("sent", nil)
+      _udp:on("receive", nil)
+      return doserver("timeout")
+    end)
+
+    -- XXX merely for diagnostics
+    _udp:on("sent", function()
+      print("sntp", "udp-on-sent")
+      _udp:on("sent", nil)
+    end)
+
+    local txts = sntppkt.make_ts(now())
+
+    _udp:on("receive", function(skt, d, port, rxip)
+      print("sntp", "udp-on-recv", rxip, port)
+
+      -- many things constitute bad packets; drop with tmr running
+      if rxip ~= ip and rxip ~= "224.0.1.1" then return end -- wrong peer
+      if port ~= 123 then return end                        -- wrong port
+      if #d   ~=  48 then return end                        -- too short
+
+      local pkt = sntppkt.proc_pkt(d, txts, now())
+
+      if pkt == nil then return -- sntppkt can also reject the packet
+      elseif type(pkt) == "string" then
+        print("sntp", "udp-on-recv", "goaway", pkt)
+        if pkt == "DENY" then -- KoD packet
+          if _kod[_six] then
+            if fk then fk("kod", serv[_six]) end
+            table.remove(_kod, _six)
+            table.remove(serv, _six)
+            _six = _six - 1 -- nextServer will add one
+          else
+            if fk then fk("goaway", serv[_six], pkt) end
+          end
+        else
+          if fk then fk("goaway", serv[_six]) end
+        end
+        return nextServer()
+      end
+
+      _kod[_six] = nil
+
+      if _pbest == serv[_six] then
+        -- this was our favorite server last time; if we don't have a
+        -- result or if we'd rather this one than the result we have...
+        if not _res or not sntppkt.pick_resp(pkt, _res, true) then
+          _res = pkt
+          _best = _pbest
+        end
+      else
+        -- this was not our favorite server; take this result if we have no
+        -- other option or if it compares favorably to the one we have, which
+        -- might be from our favorite from last pass.
+        if not _res or sntppkt.pick_resp(_res, pkt, _pbest == _best) then
+          _res = pkt
+          _best = serv[_six]
+        end
+      end
+
+      _tmr:unregister()
+      _udp:on("receive", nil)
+      _udp:on("sent", nil)
+      return nextServer()
+    end)
+
+    return _udp:send(123, ip,
+      -- '#' == 0x23: version 4, mode 3 (client), no LI
+      "#\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0"
+      .. txts)
+  end
+
+  function doserver(err)
+    if _sat == MAX_SERVER_ATTEMPTS then
+      if fk then fk(err, serv[_six]) end
+      return nextServer()
+    end
+    _sat = _sat + 1
+    local tok = _tok
+    return _udp:dns(serv[_six], function(skt, ip)
+      if tok ~= _tok then
```

Comment:
This sort of multl-level _upval_ use  can be really inscrutable.  The `tok`, `_tok` and `_udp` are all upvals to the DBS CB function, and in the case of `_udp` this is already passed as a local parameter `skt` to the function and so this should be used instead.

Reply:
Sorry, I'll call out what's going on with `tok` and `_tok` and will replace `_udp` with `skt`.

Reply:
No apology needed 😊

Reply:
Mm, I think `_tok` can go; it dates back to a time when I thought it might be worth holding on to the UDP socket between scans.  I no longer think that's likely worth doing, so it's gone, replaced with a comment.